src/OpenTelemetry.Api/Internal/Guard.cs

```diff
@@ -69,7 +69,7 @@ internal static class Guard
         /// <param name="paramName">The parameter name to use in the thrown exception.</param>
         [DebuggerHidden]
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public static void ThrowIfNull(object value, [CallerArgumentExpression("value")] string? paramName = null)
+        public static void ThrowIfNull([NotNull] object? value, [CallerArgumentExpression("value")] string? paramName = null)
```

Comment:
Both the `NotNull` and `?` is counter-intuitive. I expect this is a compiler trick, can you explain what is happening here?

Reply:
> Both the NotNull and ? is counter-intuitive. 

* `object? value` says you may pass a nullable reference into this method and it won't generate a warning.
* `[NotNull]` says that if this method returns, then the input reference was NOT `null`.

Those two things combine to make it a nice happy validation method. You may pass `null` to it, and if it returns (does not throw), then the reference is good and you won't get any more warnings about possibly accessing something which may be `null`.

You can read more about it in [this doc](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/attributes/nullable-analysis) if you are interested.

>  I expect this is a compiler trick

I would say it is more analysis tricks than compiler tricks 😄

> I also see that the other methods have the disable CS8777 but not this one.

The suppression of `CS8777` has to do with a lot of nullable magic/decorations missing in older libraries. For example, the current signature of `string.IsNullOrEmpty` is:

`public static bool IsNullOrEmpty([NotNullWhen(false)] string? value)`

The `NotNullWhen` is telling analysis about the state of the input after the method returns similar to what `NotNull` does. That decoration doesn't exist on say `net462` so you get a `CS8777` warning. Essentially because the analysis doesn't know the `IsNullOrEmpty` / `IsNullOrWhitespace` call guarantees non-null returns when `false`.